crates/sparrow-materialize/src/materialize/materialization_control.rs

```diff
@@ -0,0 +1,219 @@
+                {
+                    Ok(_) => {
+                        self.status =
+                            MaterializationProgress::Stopped(ProgressInformation::default());
+                    }
+                    Err(e) => {
+                        self.status =
+                            MaterializationProgress::Failed(ProgressInformation::default(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Running(progress) => {
+                match handle
+                    .await
+                    .into_report()
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status = MaterializationProgress::Stopped(progress.clone());
+                    }
+                    Err(e) => {
+                        self.status = MaterializationProgress::Failed(progress.clone(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Stopped(_) => {
+                tracing::info!("cannot stop; materialization already stopped")
+            }
+            MaterializationProgress::Failed(_, _) => {
+                tracing::info!("cannot stop; materialization already failed")
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Gets the status of the materialization.
+    ///
+    /// Note that progress is reported periodically, and may not contain the exact
+    /// up-to-date numbers.
+    pub fn get_status(&self) -> MaterializationStatus {
+        let progress = self.progress_receiver.borrow();
+        match &*progress {
+            MaterializationProgress::Uninitialized => MaterializationStatus {
+                id: self.id.clone(),
+                progress: ProgressInformation::default(),
+                state: State::Uninitialized,
+                error: None,
+            },
+            MaterializationProgress::Running(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Running,
+                error: None,
+            },
+            MaterializationProgress::Stopped(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Stopped,
+                error: None,
+            },
+            MaterializationProgress::Failed(progress, e) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Failed,
+                error: Some(e.to_string()),
+            },
+        }
+    }
+}
+
+/// Publish status of a materialization.
+#[derive(Debug, Default)]
+pub struct MaterializationStatus {
+    /// Unique identifier of the materialization
+    pub id: String,
```

Comment:
Why does this need the ID? If I recall, the only place this ID is used is as part of the service -- we could let that attach the ID. This, along with other suggestions then mean that `MaterializationStatus` is the same as `MaterializationProgress`, which lets you collapse those two structs and eliminate the boilerplate converting between the two.

Reply:
It doesn't - felt like more of an API convention to include the ID as part of the response, but happy to remove if it saves us that clone and we feel it's not needed.